Improve performance and reduce allocations of TemplateSegment

drieseng opened the pull request:
Improves performance and reduce allocations of **TemplateSegment(RoutePatternPathSegment other)** ctor, and adds a few unit tests.

``` ini

BenchmarkDotNet=v0.10.13, OS=Windows 10.0.17134
Intel Core i7-6700K CPU 4.00GHz (Skylake), 1 CPU, 8 logical cores and 4 physical cores
Frequency=3914069 Hz, Resolution=255.4886 ns, Timer=TSC
.NET Core SDK=2.1.400
  [Host]     : .NET Core 2.2.0-preview3-26927-02 (CoreCLR 4.6.26927.02, CoreFX 4.6.26927.02), 64bit RyuJIT
  Job-EFQGFJ : .NET Core 2.2.0-preview3-26927-02 (CoreCLR 4.6.26927.02, CoreFX 4.6.26927.02), 64bit RyuJIT

Runtime=Core  Server=True  Toolchain=.NET Core 2.2  
RunStrategy=Throughput  
```

**Before:**

|                       Method |     Mean |     Error |    StdDev |        Op/s |  Gen 0 | Allocated |
|----------------------------- |---------:|----------:|----------:|------------:|-------:|----------:|
| Ctor_RoutePatternPathSegment | 231.5 ns | 0.7288 ns | 0.6818 ns | 4,319,262.6 | 0.0057 |     536 B |
|    ToRoutePatternPathSegment | 124.3 ns | 1.5693 ns | 1.4679 ns | 8,043,307.1 | 0.0036 |     320 B |

**After:**

|                       Method |      Mean |    Error |   StdDev |         Op/s |  Gen 0 | Allocated |
|----------------------------- |----------:|---------:|---------:|-------------:|-------:|----------:|
| Ctor_RoutePatternPathSegment |  87.38 ns | 1.747 ns | 2.449 ns | 11,444,012.9 | 0.0043 |     408 B |
|    ToRoutePatternPathSegment | 124.76 ns | 1.564 ns | 1.463 ns |  8,015,433.9 | 0.0036 |     320 B |

JamesNK commented on src/Microsoft.AspNetCore.Routing/Template/TemplateSegment.cs:
```diff
@@ -18,7 +18,12 @@ public TemplateSegment()
         public TemplateSegment(RoutePatternPathSegment other)
         {
-            Parts = new List<TemplatePart>(other.Parts.Select(s => new TemplatePart(s)));
+            var partCount = other.Parts.Count;
```
This looks good but while you're here you could add:

```cs
if (other == null)
{
    throw new ArgumentNullException(nameof(other));
}
```

drieseng replied:
Done

rynowak commented on benchmarks/Microsoft.AspNetCore.Routing.Performance/Template/TemplateSegmentBenchmark.cs:
```diff
@@ -0,0 +1,38 @@
+        public void ToRoutePatternPathSegment()
+        {
+            _templateSegment.ToRoutePatternPathSegment();
+        }
```
I'm really not sure that there's any performance sensitive code here

rynowak replied:
Is this part of the code path from building a matcher?

drieseng replied:
Pretty sure it's not performance sensitive, but any performance improvement that doesn't introduce complexity should be considered. No?

That said, I'll revisit this PR later this week to see if it's still worth it now that #855 will no longer eliminate the array copy.

rynowak replied:
> Pretty sure it's not performance sensitive, but any performance improvement that doesn't introduce complexity should be considered

Sure, and believe me, I'm not complaining about getting something for free 😁 

There's a history here of people piling on this code because they understand it rather than the parts of the routing codebase that are perf-sensitive.

drieseng commented:
This is now ready for review.
Just let me know if you want me to get rid of the benchmark.

rynowak commented on test/Microsoft.AspNetCore.Routing.Tests/Template/TemplateSegmentTest.cs:
```diff
@@ -0,0 +1,46 @@
+﻿using Microsoft.AspNetCore.Routing.Patterns;
```
Needs copyright header and sort (system first)

drieseng replied:
Done

drieseng commented:
I've removed the benchmark. That should have addressed all PR feedback.

For reference, the benchmark is now available [here](https://gist.github.com/drieseng/56ab1f0bcd9902a3f44b4d1c9caf8d6e).
